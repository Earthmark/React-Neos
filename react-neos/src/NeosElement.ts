import React from "react";

interface NeosRef {}

interface NeosElement extends React.ClassAttributes<NeosRef> {
  active?: boolean;
  persistent?: boolean;
}

type ObjectVec4D = { x: number; y: number; z: number; w: number };
type ArrayVec4D = [number, number, number, number];
type SingleTVec4D = number;
export type Vec4D = ObjectVec4D | ArrayVec4D | SingleTVec4D;

type ObjectVec3D = { x: number; y: number; z: number };
type ArrayVec3D = [number, number, number];
type SingleTVec3D = number;
export type Vec3D = ObjectVec3D | ArrayVec3D | SingleTVec3D;

type ObjectVec2D = { x: number; y: number };
type ArrayVec2D = [number, number];
type SingleTVec2D = number;
export type Vec2D = ObjectVec2D | ArrayVec2D | SingleTVec2D;

type FixedColor = "red";
// A single number is interpreted as a vec4.
export type Color = Vec4D | Vec3D | FixedColor;

type DetailedNeos3DElementProps = NeosElement & {
  position?: Vec3D;
  rotation?: Vec3D;
  scale?: Vec3D;
};
type DetailedNeos2DElementProps = NeosElement & {
  anchorMin?: Vec2D;
  anchorMax?: Vec2D;
  offsetMin?: Vec2D;
  offsetMax?: Vec2D;
  pivot?: Vec2D;
};

export interface NeosElements {
  nSlot: DetailedNeos3DElementProps & {
    children?: DetailedNeos3DElementProps[] | DetailedNeos3DElementProps;
  };
  nSmoothTransform: DetailedNeos3DElementProps & {
    children?: DetailedNeos3DElementProps[] | DetailedNeos3DElementProps;
  };
  nCanvas: DetailedNeos3DElementProps & {
    size?: Vec2D;
    children?: DetailedNeos2DElementProps[] | DetailedNeos2DElementProps;
  };

  nRectTransform: DetailedNeos2DElementProps & {
    children?: DetailedNeos2DElementProps[] | DetailedNeos2DElementProps;
  };
  nText: DetailedNeos2DElementProps & {
    children?: string;
    nullText?: string;
    parseRichText?: boolean;
    horizontalAlignment?: "left" | "center" | "right" | "justify";
    color?: Color;
    size?: number;
    horizontalAutoSize?: boolean;
    verticalAutoSize?: boolean;
  };
}

declare global {
  namespace JSX {
    interface IntrinsicElements extends NeosElements {}
  }
}
