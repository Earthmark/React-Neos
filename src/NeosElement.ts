import React, { ReactElement } from "react";
import { PropsDelta } from "./NeosRenderer";
import { differs, DiffFunc } from "./Primitives";

function HasReactChildren(): {
  children: DiffFunc<ReactElement | Array<ReactElement>>;
} {
  return {
    children: () => null,
  };
}

const baseElementDef = {
  active: differs.bool,
  persistent: differs.bool,
  name: differs.string,
  tag: differs.string,
};

const base3DElementDef = {
  ...baseElementDef,
  position: differs.float3,
  rotation: differs.floatQ,
  scale: differs.float3,
};

const base2DElementDef = {
  ...baseElementDef,
  anchorMin: differs.float2,
  anchorMax: differs.float2,
  offsetMin: differs.float2,
  offsetMax: differs.float2,
  pivot: differs.float2,
};

export const elementDefs = mapDefsToDiffers({
  transform: {
    ...base3DElementDef,
    ...HasReactChildren(),
  },
  smoothTransform: {
    ...base3DElementDef,
    ...HasReactChildren(),
    smoothTransformEnabled: differs.bool,
    smoothSpeed: differs.float,
  },
  spinner: {
    ...base3DElementDef,
    ...HasReactChildren(),
    speed: differs.float3,
    range: differs.float3,
  },
  box: {
    ...base3DElementDef,
    albedoColor: differs.color,
    emissiveColor: differs.color,
    size: differs.float3,
    colliderActive: differs.bool,
    characterCollider: differs.bool,
    ignoreRaycasts: differs.bool,
  },
  canvas: {
    ...base3DElementDef,
  },
  rect: {
    ...base2DElementDef,
  },
  text: {
    ...base2DElementDef,
    children: differs.string,
  },
  button: {
    ...base2DElementDef,
  },
});

function defsToDiffer<Props>(
  elementDef: PropsDef<Props>
): PropsDiffer<Partial<Props>> {
  return (a, b, d) => {
    for (const key in elementDef) {
      const differ = elementDef[key];
      const result = differ(a[key], b[key]);
      if (result !== null) {
        d.propDiffs.push({
          ...result,
          prop: key,
        });
      }
    }
  };
}

export function mapDefsToDiffers<PropMap>(o: {
  [Type in keyof PropMap]: PropsDef<PropMap[Type]>;
}): {
  [Type in keyof PropMap]: PropsDiffer<Partial<PropMap[Type]>>;
} {
  const result: any = {};
  for (const key in o) {
    result[key] = defsToDiffer(o[key]);
  }
  return result;
}

const elementProps: {
  [Type in keyof typeof elementDefs]: (
    p: PropsFromDiffer<typeof elementDefs[Type]>
  ) => React.ReactElement<PropsFromDiffer<typeof elementDefs[Type]>, Type>;
} = Object.fromEntries(
  Object.keys(elementDefs).map((e) => [e, e])
) as unknown as any;

type PropsDiffer<Props> = (a: Props, b: Props, d: PropsDelta) => void;

type PropsDef<Props> = {
  [Prop in keyof Props]: DiffFunc<Props[Prop]>;
};

type PropsFromDiffer<Def> = Def extends PropsDiffer<infer Props>
  ? Props
  : never;

export type ElementProps = {
  [P in keyof typeof elementDefs]: PropsFromDiffer<typeof elementDefs[P]>;
};

declare global {
  namespace JSX {
    interface IntrinsicElements extends ElementProps {}
  }
}

export default elementProps;
