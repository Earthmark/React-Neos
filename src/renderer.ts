import React from "react";
import Reconciler from "react-reconciler";
import { performance } from "perf_hooks";
import { ElementId, InboundSignal, OutboundSignal, PropUpdate } from "./signal.js";
import { componentDefs } from "./components.js";

export interface ElementTemplate<Props, Refs> {
  updater: ElementUpdater<Props>;
  refFactory: ElementRefFactory<Refs>;
}

export type ElementUpdater<Props = {}> = (
  oldProps: Partial<Props>,
  newProps: Partial<Props>,
  update: {
    diff: (prop: PropUpdate) => void;
    //register: (handler: ((arg: string) => void) | undefined) => () => void;
  }
) => void;

export type ElementRefFactory<Refs> = (id: ElementId) => FieldRefs<Refs>;

export type FieldRefs<FieldTypes> = {
  [Field in keyof FieldTypes]: FieldRef<Extract<FieldTypes[Field], string>>;
};

export interface FieldRef<TypeName extends string> {
  type: TypeName;
  name: string;
  elementId: string;
}

export interface ReactNeosRenderer {
  createInstance(handler: (signal: OutboundSignal) => void): {
    render(signal: InboundSignal): void;
  };
}

interface Container extends LogicalInstance {
  globalInstanceId: number;
  globalEventId: number;
}

interface RenderedInstance extends LogicalInstance {
  container: Container;
  updater: ElementUpdater<any>;
  refs: FieldRefs<any>;
}

interface LogicalInstance {
  id: string;
  children: Record<string, LogicalInstance>;
  events: Record<string, (arg: string) => void>;
}

function getElementById(
  id: string,
  instance: LogicalInstance
): LogicalInstance | undefined {
  if (instance.id === id) {
    return instance;
  }
  for (const child of Object.values(instance.children ?? {})) {
    const result = getElementById(id, child);
    if (result) {
      return result;
    }
  }
}

export default function createRender<
  AdditionalComponents extends Record<
    keyof AdditionalComponents,
    ElementTemplate<any, any>
  >
>(
  node: React.ReactNode,
  elementTemplates?: AdditionalComponents
): ReactNeosRenderer {
  return {
    createInstance(handler: (signal: OutboundSignal) => void) {
      const components = elementTemplates ?? componentDefs;
      const reconciler = Reconciler<
        Extract<keyof typeof components, string>,
        Record<string, any>,
        Container,
        RenderedInstance,
        never,
        never,
        never,
        FieldRefs<any>,
        {},
        {
          diffs: Array<PropUpdate>;
        },
        never,
        NodeJS.Timer,
        number
      >({
        supportsMutation: true,
        supportsHydration: false,
        supportsPersistence: false,
        noTimeout: -1,
        isPrimaryRenderer: true,

        prepareUpdate(instance, type, oldProps, newProps) {
          const diffs: Array<PropUpdate> = [];
          instance.updater(oldProps, newProps, {
            diff: (prop) => {
              diffs.push(prop);
            },
            // register: (handler) => {
            //   return () => {};
            // },
          });
          return diffs.length > 0 ? { diffs } : null;
        },

        createInstance(type, props, container, context) {
          const id = `${container.globalInstanceId++}`;
          handler({
            signal: "create",
            id,
            type,
          });
          const template = components[type];
          if (template === undefined) {
            throw new Error(`Unknown element type ${type}`);
          }
          const instance = {
            id,
            container,
            updater: template.updater,
            refs: template.refFactory(id),
            children: {},
            events: {},
          };

          const diffs: Array<PropUpdate> = [];
          instance.updater({}, props as any, {
            diff: (prop) => {
              diffs.push(prop);
            },
            // register: (handler) => {
            //   return () => {};
            // },
          });
          const delta = diffs.length > 0 ? { diffs } : null;

          if (delta && delta.diffs.length > 0) {
            handler({
              signal: "update",
              id,
              props: delta.diffs,
            });
          }
          return {
            id,
            container,
            updater: template.updater,
            refs: template.refFactory(id),
            children: {},
            events: {},
          };
        },
        createTextInstance() {
          throw new Error(
            "Manually setting text isn't supported, wrap it in a text element."
          );
        },

        clearContainer(container) {
          container.children = {};
          handler({
            signal: "remove",
            id: container.id,
          });
        },

        appendInitialChild(parentInstance, child) {
          parentInstance.children[child.id] = child;
          handler({
            signal: "setParent",
            id: child.id,
            parentId: parentInstance.id,
          });
        },

        appendChild(parentInstance, child) {
          parentInstance.children[child.id] = child;
          handler({
            signal: "setParent",
            id: child.id,
            parentId: parentInstance.id,
          });
        },
        appendChildToContainer(container, child) {
          container.children[child.id] = child;
          handler({
            signal: "setParent",
            id: child.id,
            parentId: container.id,
          });
        },

        insertBefore(parentInstance, child, beforeChild) {
          parentInstance.children[child.id] = child;
          handler({
            signal: "setParent",
            id: child.id,
            parentId: parentInstance.id,
            after: beforeChild.id,
          });
        },
        insertInContainerBefore(container, child, beforeChild) {
          container.children[child.id] = child;
          handler({
            signal: "setParent",
            id: child.id,
            parentId: container.id,
            after: beforeChild.id,
          });
        },

        removeChild(parentInstance, child) {
          delete parentInstance.children[child.id];
          handler({
            signal: "remove",
            id: child.id,
          });
        },
        removeChildFromContainer(container, child) {
          delete container.children[child.id];
          handler({
            signal: "remove",
            id: child.id,
          });
        },

        finalizeInitialChildren() {
          return false;
        },

        commitUpdate(instance, updatePayload) {
          handler({
            signal: "update",
            id: instance.id,
            props: updatePayload.diffs,
          });
        },

        shouldSetTextContent(type) {
          return type === "text";
        },

        getRootHostContext() {
          return {};
        },
        getChildHostContext() {
          return {};
        },

        getPublicInstance(instance) {
          return instance.refs;
        },
        prepareForCommit() {
          return null;
        },
        resetAfterCommit() {},
        preparePortalMount() {},

        now: performance.now,
        scheduleTimeout: setTimeout,
        cancelTimeout: clearTimeout,
      });
      const containerInfo: Container = {
        id: "root",
        globalInstanceId: 1,
        globalEventId: 1,
        children: {},
        events: {},
      };
      const container = reconciler.createContainer(
        containerInfo,
        0,
        null,
        false,
        null,
        "rn",
        (e) => {},
        null
      );

      reconciler.updateContainer(node, container);

      return {
        render(signal: InboundSignal): void {
          const instance = getElementById(signal.id, container);
          if (instance) {
            const event = instance.events[signal.signal];
            if (event) {
              event(signal.arg);
            }
          }
        },
      };
    },
  };
}
