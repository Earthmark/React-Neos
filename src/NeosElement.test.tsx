import React from 'react';
import renderer from "./NeosRenderer";
import TestRenderer from "react-test-renderer";
import {ElementPropStringifyMap} from "./NeosElement";

test("Verify failure", () => {
  expect(renderer).toBeDefined();
  expect(true).toBe(true);
});

test("Verify hierarchy shows as expected", () => {
  expect(TestRenderer.create(<nTransform scale={3} />).toJSON()).toMatchSnapshot();
});

test("verify slot stringifies as expected", () => {
  const dest: Array<string> = [];
  ElementPropStringifyMap.nTransform(
    {
    oldProps: {
      active: true,
      persistent: true,
      scale: 2,
    },
    newProps: {
      active: null,
      persistent: true,
      scale: 3,
    },
    arr: dest,
    events: {}
  });
  expect(dest).toStrictEqual(["active=bool#$", "scale=float3#[3;3;3]"]);
});

type TestCaseMap = {
  [key in keyof typeof ElementPropStringifyMap]: 
    PropsOfStringify<typeof ElementPropStringifyMap[key]>[]
};

type PropsOfStringify<G> = G extends (g: infer Gatherer) => unknown
  ? Gatherer
  : never;

const testCases : TestCaseMap = {
  nTransform: [{
    oldProps: {
      active: true,
      persistent: true,
      scale: 2,
    },
    newProps: {
      active: null,
      persistent: true,
      scale: 3,
    },
    arr: ["active=bool#$", "scale=float3#[3;3;3]"],
    events: {}
  },
  {
    oldProps: {
    },
    newProps: {
      active: null,
      persistent: true,
      position: {x: 1, y: 2, z: 43},
      scale: 3,
    },
    arr: [
      "persistent=bool#true",
      "position=float3#[1;2;43]",
      "scale=float3#[3;3;3]"
    ],
    events: {}
  }],
  nSmoothTransform: [],
  nCanvas: [],
  nRectTransform: [],
  nText: [],
  nButton: []
}

it.each(
  Object.entries(testCases).flatMap(([k, v]) =>
    v.map((o) => ({ name: k, ...o}))
  )
)("element processes expected diff for set %s", ({name, oldProps, newProps, arr}) => {
  const src: Array<string> = [];
  (ElementPropStringifyMap as any)[name as any]({
    oldProps,
    newProps,
    arr: src
  });
  expect(src).toStrictEqual(arr);
});

