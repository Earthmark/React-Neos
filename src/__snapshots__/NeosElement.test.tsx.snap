// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify hierarchy shows as expected 1`] = `
<nTransform
  scale={3}
/>
`;
